lib/android/fido/state.dart

```diff
@@ -165,26 +166,152 @@ final androidFingerprintProvider = AsyncNotifierProvider.autoDispose
         _FidoFingerprintsNotifier.new);
 
 class _FidoFingerprintsNotifier extends FidoFingerprintsNotifier {
+  final _events = const EventChannel('android.fido.fingerprints');
+  late StreamSubscription _sub;
+
   @override
   FutureOr<List<Fingerprint>> build(DevicePath devicePath) async {
-    return [];
+    _sub = _events.receiveBroadcastStream().listen((event) {
+      final json = jsonDecode(event);
+      if (json == null) {
+        state = const AsyncValue.loading();
+      } else {
+        List<Fingerprint> newState = List.from(
+            (json as List).map((e) => Fingerprint.fromJson(e)).toList());
+        state = AsyncValue.data(newState);
+      }
+    }, onError: (err, stackTrace) {
+      state = AsyncValue.error(err, stackTrace);
+    });
+
+    ref.onDispose(_sub.cancel);
+    return Completer<List<Fingerprint>>().future;
   }
 
   @override
   Stream<FingerprintEvent> registerFingerprint({String? name}) {
     final controller = StreamController<FingerprintEvent>();
+    const registerEvents = EventChannel('android.fido.registerFp');
+
+    final registerFpSub =
+        registerEvents.receiveBroadcastStream().skip(1).listen((event) {
+      if (controller.isClosed) {
+        _log.debug('Controller already closed, ignoring: $event');
+      }
+      _log.debug('Received register fingerprint event: $event');
+      if (event is String && event.isNotEmpty) {
+        final e = jsonDecode(event);
+        _log.debug('Received register fingerprint event: $e');
+
+        final status = e['status'];
+
+        controller.sink.add(switch (status) {
+          'capture' => FingerprintEvent.capture(e['remaining']),
+          'capture-error' => FingerprintEvent.error(e['code']),
+          final other => throw UnimplementedError(other)
+        });
+      }
+    });
+
+    controller.onCancel = () async {
+      if (!controller.isClosed) {
+        _log.debug('Cancelling fingerprint registration');
+        await _methods.invokeMethod('cancelRegisterFingerprint');
+        await registerFpSub.cancel();
+      }
+    };
+
+    controller.onListen = () async {
+      try {
+        final registerFpResult =
+            await _methods.invokeMethod('registerFingerprint', {'name': name});
+
+        _log.debug('Finished registerFingerprint with: $registerFpResult');
+
+        final resultJson = jsonDecode(registerFpResult);
+
+        if (resultJson['success'] == true) {
+          controller.sink
+              .add(FingerprintEvent.complete(Fingerprint.fromJson(resultJson)));
+        } else {
+          // TODO abstract platform errors
+          final errorStatus = resultJson['status'];
+          if (errorStatus != 'user-cancelled') {
+            throw RpcError(errorStatus, 'Platform error: $errorStatus', {});
+          }
+        }
+
+        await controller.sink.close();
```

Comment:
Should we close in a `finally`-clause to avoid leaving the steam open if there is an exception?

Reply:
fixed